---
title: "RAD.reference.aligned.vignette"
author: "Devon DeRaad"
date: "8/24/2020"
output: html_document
---

```{r setup}
library(vcfR)
library(adegenet)
library(adegraphics)
library(pegas)
library(StAMPP)
library(lattice)
library(gplots)
library(ape)
library(ggmap) 
library(ggplot2)
```


#Hard filters can be applied immediately, to ensure that we are only considering SNPs of reasonable quality when we have to make subjective decisions downstream about which samples to keep and where to set arbitrary filtering cutoffs:
#Filters that use % or quantile based cutoffs should not be applied until after we have decided which individuals must be dropped, because changing our sample will shift these percentage and quantile based cutoffs.




#chromR visualization to ensure that hard quality filtering was successful and we don't see an abundance of paralogs, low depth variants, etc.
```{r chromR}
summary(cars)
```


#Now we can visualize missing data and depth of sequencing by sample and by population
#Determining which samples to retain is highly project specific, and is contingent on your sampling, your taxa, the sequencing results, etc.
#It is also wise to take missing data into account by population. Often we don't know a-priori what our populations will look like, but in general we want to avoid making inferences about populations if they have consistently less information than the rest of our dataset.
#On the flip side of that coin, you may have designed a project to get information about a rare sample/population that lacks existing genetic sampling, and therefore must retain specific samples despite low sequencing and high missing data. This is a reality, and simply needs to be addressed carefully.

```{r}

```


#Now that we have our final set of samples for downstream analysis, we can run additional quality filtering based on percentages and quantiles, to further increase our signal to noise ratio
```{r}

```


#Now we must set the arbitrary missing data cutoff
#We can visualize the effect that typical missing data cutoffs will have on both the number of SNPs retained and the total missing data in our entire dataset
#We want to choose a cutoff that minimizes the overall missing data in the dataset, while maximizing the total number of loci retained
#This will also depend on the samples that we decided to include above. A good rule of thumb is that samples shouldn't be above 50% missing data after applying this cutoff. So if we are retaining low data samples out of necessity or project design, we may have to set a more stringent cutoff at the expense of total SNPs retained for downstream analyses.
```{r}

```


#Now, we should investigate the effect of a minor allele frequency cutoff on our downstream inferences. MAF cutoffs can be helpful in removing spurious and uninformative loci from the dataset, but also have the potential to bias downstream inferences. Linck and Battey (2019) have an excellent paper on just this topic.
#This package contains some convenient wrapper functions that streamline the investigation of MAF cutoffs.
#Once again, implementing a MAF cutoff is going to depend on your project and the amount of divergence between your populations. For something like phylogenetic reconstruction of deeply diverged species, a MAF cutoff may increase signal to noise ratio and make your SNP dataset more computationally tractible. In contrast, your ability to identify fine scale geographic variation between sampling localities that are actively exchanging migrants may be hampered by a MAF cutoff as recently derived low frequency, private alleles may be the only thing separating these populations.
```{r}

```


#Finally, you can use this filtered vcf file directly for downstream analyses, or you can filter to retain only physically unlinked loci (which is a requirement of some programs, e.g. structure). This step should always be done last, because it is not quality aware. To introduce a quality-blind linkage filter before doing the quality based filtering here would potentially remove quality SNPs while retaining low quality SNPs in a single locus or region.
#If filtering for linkage is needed, it can be done with a simple one-liner via vcftools (set thin to whatever bp difference you assume linakge decays at in your study organism)
#vcftools --vcf vcf.vcf --thin 10000 --recode --out vcf


